File: day3/day3.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func countTrees(right int, down int) int {
	trees := 0
	file, _ := os.Open("day3.in")

	currentPos := 0
	scanner := bufio.NewScanner(file)
	for i := 0; scanner.Scan(); i++ {
		if down < 2 && i == 0 {
			continue
		}
		if i%down != 0 {
			continue
		}

		line := scanner.Text()
		currentPos = (currentPos + right) % len(line)
		if line[currentPos] == '#' {
			trees++
		}
	}
	return trees
}

func main() {
	trees11 := countTrees(1, 1)
	trees31 := countTrees(3, 1)
	trees51 := countTrees(5, 1)
	trees71 := countTrees(7, 1)
	trees12 := countTrees(1, 2)
	answer := trees11 * trees31 * trees51 * trees71 * trees12
	fmt.Println(trees12)
	fmt.Println(answer)
}

File: day3/day3.py
def count_trees(right, down):
    trees = 0
    with open('day3.in') as file:
        if down < 2:
            first_line = file.readline()
        current_pos = 0
        for line in file.read().split("\n")[::down]:
            current_pos = (current_pos + right) % len(line)
            if line[current_pos] == '#':
                trees += 1
    return trees

def main():
    trees11 = count_trees(1, 1)
    trees31 = count_trees(3, 1)
    trees51 = count_trees(5, 1)
    trees71 = count_trees(7, 1)
    trees12 = count_trees(1, 2)
    answer  = trees11 * trees31 * trees51 * trees71 * trees12
    print(answer)

main()